### PR TITLE
Add TLS tests and sync.pool for bridge

### PR DESCRIPTION
**Pull request by natnaelg**
Add BenchmarkFullProxyTLS and TestFullProxyTLS to proxy_test.go

Change proxy.bridge to use a sync.pool to reduce memory allocs

Before & after

```
Running tool: /usr/local/go/bin/go test -benchmem -run=^$ github.com/nccgroup/tracy/proxy -bench ^BenchmarkFullProxyTLS$

goos: darwin
goarch: amd64
pkg: github.com/nccgroup/tracy/proxy
/Users/nati/golang/src/github.com/nccgroup/tracy/proxy/BenchmarkFullProxyTLS-4   	[ERROR]:2018/07/21 10:50:59 proxy.go:61: too many SQL variables
     500	   2961541 ns/op	 1127899 B/op	   16400 allocs/op
PASS
ok  	github.com/nccgroup/tracy/proxy	1.832s

```
```

Running tool: /usr/local/go/bin/go test -benchmem -run=^$ github.com/nccgroup/tracy/proxy -bench ^BenchmarkFullProxyTLS$

goos: darwin
goarch: amd64
pkg: github.com/nccgroup/tracy/proxy
BenchmarkFullProxyTLS-4   	     500	   2939474 ns/op	 1129618 B/op	   16377 allocs/op
PASS
ok  	github.com/nccgroup/tracy/proxy	1.819s
Success: Benchmarks passed.

```

### CODE REVIEW

**Comment by heathj** [comment]
I'm curious what that error is: `[ERROR]:2018/07/21 10:50:59 proxy.go:61: too many SQL variables`

**Comment by natnaelg** [comment]
Yea, it happens when I run the test too much, goes away when I delete my database and start from scratch. 🤷‍♂️

File: proxy/proxy.go [hunk 1]
```diff
@@ -396,26 +403,9 @@ func handleConnection(client net.Conn) {
...
-		}
-
-		// Copy the bytes read above to the other connection.
-		_, err = server.Write(buf[:nb])
-
-		if err != nil {
-			log.Error.Println(err)
-			break
-		}
+	buf := bufferPool.Get().([]byte)
```
Comment (heathj) [inline proxy/proxy.go]: 👍

File: proxy/proxy.go [hunk 2]
```diff
@@ -396,26 +403,9 @@ func handleConnection(client net.Conn) {
...
+	if _, err := io.CopyBuffer(client, server, buf); err != nil {
```
Comment (heathj) [inline proxy/proxy.go]: This is much nicer, thanks.

**Comment by heathj** [comment]
Maybe this is the cause: https://github.com/jinzhu/gorm/issues/1054

**Comment by heathj** [comment]
@natnaelg Looks like the for loop in the bridge function got deleted in the merge, sorry should have caught that. I just pushed the fix; might be good to rerun your benchmarks. I bet you get better performance.

**Comment by heathj** [comment]
Also, bridge is only called for websockets, so might be good to start thinking how you will test websocket connections. Maybe we can steal from https://github.com/gorilla/websocket/blob/master/client_server_test.go

**Comment by natnaelg** [comment]
@heathj That was intentional, the io.CopyBuffer handles the loop until it finds an EOF, did you test it and it broke?

https://github.com/golang/go/blob/master/src/io/io.go#L401-L423

**Comment by heathj** [comment]
O cool. Hmm, yea I got an error that said something about writing or reading from a closed connection. I'll play with it and give more detailed information.

**Comment by heathj** [comment]
I'll revert my changes until I lock it down, if it was anything at all. Thanks for showing me that, that's cool.

**Comment by natnaelg** [comment]
Did the error have a line associated with it?

**Comment by natnaelg** [comment]
This might be from the race conditions, one go routine is passed a request but the main routine has finished and has closed the conn

**Comment by natnaelg** [comment]
That can actually be fixed by adding a sync.WaitGroup, it won't fix the race condition but it'll make sure you don't close the connection while others are using it. I'll make a pr.

**Pull request merged**